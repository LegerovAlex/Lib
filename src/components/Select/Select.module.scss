.wrapper {
  position: relative;
  display: inline-flex;
  flex-direction: column;

  &:focus-within .label {
    color: #1976d2;
  }
}

.label {
  position: absolute;
  top: -8px;
  left: 12px;
  padding: 0 4px;
  background-color: #fff;
  font-size: 14px;
  color: #5a5a5a;
}

.select {
  display: flex;
  justify-content: space-between;
  align-items: center;
  min-width: 160px;
  gap: 20px;
  border: 1px solid #5a5a5a;
  border-radius: 4px;
  padding: 16px 12px 16px;
  cursor: pointer;
  background-color: #fff;
  transition: border-color 0.2s ease;

  &:hover {
    border-color: #007bff;
  }

  &:focus {
    border-color: #007bff;
  }
}

.value {
  color: #222;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.arrow {
  font-size: 12px;
  color: #666;
}

.menu {
  position: absolute;
  top: calc(100% + 4px);
  left: 0;
  right: 0;
  max-height: 200px;
  overflow-y: auto;
  background-color: white;
  border: 1.5px solid #ccc;
  border-radius: 4px;
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.1);
  padding: 0;
  margin: 0;
}

.disabled {
  background: #f5f5f5;
  color: #aaa;
  cursor: not-allowed;
  pointer-events: none;
}
