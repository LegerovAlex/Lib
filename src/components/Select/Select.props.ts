export interface SelectOptions {
  id: number;
  value: string;
  label: string;
}

export interface SelectProps {
  label?: string;
  value: string;
  options: SelectOptions[];
  onChange: (value: string) => void;
  placeholder?: string;
  disabled?: boolean;
}
