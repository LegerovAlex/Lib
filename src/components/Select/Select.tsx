import { FC, useEffect, useRef, useState } from 'react';
import { SelectProps } from './Select.props';
import styles from './Select.module.scss';
import { MenuItem } from './MenuItem/MenuItem';

export const Select: FC<SelectProps> = ({
  label,
  value,
  onChange,
  placeholder,
  disabled,
  options,
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const selectRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (selectRef.current && !selectRef.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    };
    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  const handleToggle = () => {
    if (disabled) return;
    setIsOpen((prev) => !prev);
  };

  const handleSelect = (selectedValue: string) => {
    onChange(selectedValue);
    setIsOpen(false);
  };

  const selectedOption = options.find((opt) => opt.value === value);

  return (
    <div className={styles.wrapper} ref={selectRef}>
      {label && <label className={styles.label}>{label}</label>}
      <div
        role="combobox"
        tabIndex={0}
        className={`${styles.select} ${disabled && styles.disabled}`}
        onClick={handleToggle}
      >
        <span className={styles.value}> {selectedOption ? selectedOption.label : placeholder}</span>
        <span className={styles.arrow}>â–¼</span>
      </div>
      {isOpen && (
        <ul className={styles.menu}>
          {options.map((opt) => (
            <MenuItem disabled={disabled} key={opt.id} value={opt.value} onSelect={handleSelect}>
              {opt.label}
            </MenuItem>
          ))}
        </ul>
      )}
    </div>
  );
};
