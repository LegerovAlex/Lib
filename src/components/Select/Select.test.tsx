import { render, screen } from '@testing-library/react';
import { SelectProps } from './Select.props';
import { Select } from './Select';
import userEvent from '@testing-library/user-event';

export interface RenderSelectResult {
  select: HTMLElement;
  onChange: jest.Mock;
}

export function renderSelect(props: Partial<SelectProps> = {}): RenderSelectResult {
  const onChange = jest.fn();

  const defaultProps: SelectProps = {
    label: 'Label',
    value: '',
    onChange,
    placeholder: 'Select...',
    disabled: false,
    options: [
      { id: 1, value: 'one', label: 'Option One' },
      { id: 2, value: 'two', label: 'Option Two' },
    ],
    ...props,
  };

  render(<Select {...defaultProps} />);

  const select = screen.getByRole('combobox');

  return { onChange, select };
}

describe('Select', () => {
  it('renders with placeholder when no value is selected', () => {
    const { select } = renderSelect();
    expect(select).toBeInTheDocument();
    expect(select).toHaveTextContent('Select...');
  });
  it('renders option label when value is provided', () => {
    const { select } = renderSelect({ value: 'one' });
    expect(select).toBeInTheDocument();
    expect(select).toHaveTextContent('Option One');
  });
  it('renders label when  is provided', () => {
    renderSelect({ label: 'Test Label' });
    const lable = screen.getByText('Test Label');
    expect(lable).toBeInTheDocument();
  });
  it('toggles menu when on click', async () => {
    const { select } = renderSelect();
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
    await userEvent.click(select);
    expect(screen.getByRole('list')).toBeInTheDocument();
    await userEvent.click(select);
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });
  it('calls onChange with value when options is selected ', async () => {
    const { select, onChange } = renderSelect();
    await userEvent.click(select);
    const option = screen.getByText('Option One');
    await userEvent.click(option);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('one');
  });
  it('does not call onChange when disabled is provided', async () => {
    const { select, onChange } = renderSelect({ disabled: true });
    await userEvent.click(select);
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
    expect(onChange).not.toHaveBeenCalled();
  });
  it('closes menu when clicking outside', async () => {
    const { select } = renderSelect();
    await userEvent.click(select);
    expect(screen.getByRole('list')).toBeInTheDocument();

    await userEvent.click(document.body);
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });
  it('matches snapshot when enabled', async () => {
    const { select } = renderSelect();
    await userEvent.click(select);
    expect(screen.getByRole('list')).toMatchSnapshot();
  });
  it('matches snapshot when disabled', () => {
    const { select } = renderSelect({ disabled: true });
    expect(select).toMatchSnapshot();
  });
});
