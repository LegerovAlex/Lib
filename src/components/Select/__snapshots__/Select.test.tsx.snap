// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Select matches snapshot when disabled 1`] = `
<div
  class="select disabled"
  role="combobox"
  tabindex="0"
>
  <span
    class="value"
  >
     
    Select...
  </span>
  <span
    class="arrow"
  >
    â–¼
  </span>
</div>
`;

exports[`Select matches snapshot when enabled 1`] = `
<ul
  class="menu"
>
  <li
    class="menuItem false"
  >
    Option One
  </li>
  <li
    class="menuItem false"
  >
    Option Two
  </li>
</ul>
`;
